### PR TITLE
[NoQA] feat: react-compiler

### PR DESCRIPTION
**Pull request by kirillzyusko**
### Details

Added `react-compiler` that automatically applies necessary `memo`/`useMemo`/etc. and improves performance of the app.

To add this compiler I had:
- to patch `eslint` plugin (the PR was merged https://github.com/facebook/react/pull/29104 -> will be available in next release);
- to add `react-compiler-runtime` package manually (the `react-compiler-runtime` is available by default in `react@19`, but `react@19` is not compatible with RN 0.73/0.74, so we had to wait before the upgrade - more reasonable was to add missing runtime and link it manually for now);

Below two screenshots from `react-devtools` showing that compiler actually works:

|Before|After|
|-------|-----|
|image|image|

> [!WARNING]
> As of `react-native@0.73` the new badge with memo will not be shown (though optimization will be applied. Read an [explanation](https://github.com/Expensify/App/issues/42211#issuecomment-2117760659) here.

After running performance e2e tests I got next results (comparing to current `main`):

❇️ Performance comparison results:
➡️ Significant changes to duration

Open Chat Finder Page TTI: 2758.066 ms → 2639.980 ms (-118.086 ms, -4.3%)
➡️ Meaningless changes to duration

App start nativeLaunch: 91.750 ms → 93.517 ms (+1.767 ms, +1.9%)
App start nativeLaunchEnd_To_appCreationStart: 91.737 ms → 90.877 ms (-0.860 ms, -0.9%)
App start appCreation: 108.133 ms → 111.817 ms (+3.683 ms, +3.4%)
App start appCreationEnd_To_contentAppeared: 1116.733 ms → 1108.483 ms (-8.250 ms, -0.7%)
App start contentAppeared_To_screenTTI: 2324.047 ms → 2343.516 ms (+19.469 ms, +0.8%)
App start runJsBundle: 863.983 ms → 849.467 ms (-14.517 ms, -1.7%)
App start TTI: 3745.847 ms → 3750.249 ms (+4.403 ms, ±0.0%)
App start regularAppStart: 0.060 ms → 0.061 ms (+0.000 ms, ±0.0%)
Load Search Options: 403.670 ms → 415.041 ms (+11.370 ms, +2.8%)
Chat opening: 84.929 ms → 87.507 ms (+2.578 ms, +3.0%)
Chat TTI: 1522.559 ms → 1513.895 ms (-8.664 ms, -0.6%)

### Fixed Issues

$ https://github.com/Expensify/App/issues/42211
PROPOSAL: N/A

### Tests

- [ ] Verify that no errors appear in the JS console

### Offline tests

N/A

### QA Steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Review by Ollyws — APPROVED** [review]
LGTM.

**Comment by kirillzyusko** [comment]
@mountiny what are the next steps here? I remember you wanted to discuss it on Monday so it's a friendly reminder 👀 

My understanding is that it's pretty stable and we have a control of each file (so we always can disable unnecessary optimizations if needed).

**Review by mountiny — DISMISSED** [review]
thank you @kirillzyusko this seems like it works and tests well

I am curious if @roryabraham @AndrewGable have any concerns with adding this now

**Comment by AndrewGable** [comment]
Should we run an AdHoc regression test to see any thing breaks? It seems like it might be a decent idea

**Comment by mountiny** [comment]
running

**Comment by roryabraham** [comment]
Also, does this mean that we should start changing the way we write code such that we don't use `useMemo`, `useCallback`, etc... and just let the React Compiler handle memoization on its own?

**Comment by roryabraham** [comment]
I feel like the real test here would be to take some optimized but complex/less-readable code in an important component, remove all memoization, and then see how well the React Compiler does to automatically add memoization.

**Comment by mountiny** [comment]
Asked for QA testing https://expensify.slack.com/archives/C9YU7BX5M/p1717521040631869

**Comment by kirillzyusko** [comment]
> There are no changes for the failing test on main, can you see if the npm run gh-actions-build command does produce a diff on this branch? I think we should commit those changes if so.

@AndrewGable I pulled lates main again, and it produced a different output again. I committed modified file, but may I ask you to shed some light on why it happens? I simply added new dependency to `package.json` - why it affected a compilation of GH action file? 🤔

**Comment by kirillzyusko** [comment]
> Should we run an AdHoc regression test to see any thing breaks? It seems like it might be a decent idea

@mountiny may I kindly ask you to re-trigger a new build? I think for a previous build I haven't updated a branch for a pretty long time, so I assume it may have old bugs. Now I rebased all my changes, so it should be in more up-to-date state (but if you think that we can test previous build, then I'm not against it).

**Comment by mountiny** [comment]
New build runnig here https://github.com/Expensify/App/actions/runs/9439164596

**Comment by kirillzyusko** [comment]
@mountiny is there any updates from QA team?..

**Comment by mountiny** [comment]
@kirillzyusko I think they had lots of other builds and they were not able to prioritize this one. Do you think you could sync with main and we can create a fresh build for them. Sorry that this is taking longer

**Comment by kavimuru** [comment]
@mountiny we just started testing on this build

**Comment by lanitochka17** [comment]
##  Setting - App crashes after click on "Settings > Share Code"

**Version Number:** 1.4.78-4
**Reproducible in staging?:** N
**Reproducible in production?:** N
**Issue reported by:** Applause-Internal team


## Action Performed:
1. Open Web or iOS App / 42287 Special Build
2. Login with any account
3. Go to Settings
4. Click / Tap on "Share Code" button

## Expected Result:
"Share Code" page is оpened

## Actual Result:
App crashes after click on "Settings > Share Code"

## Workaround:
Unknown


## Platforms:

Which of our officially supported platforms is this issue occurring on?

- [ ] iOS: Native
- [x] MacOS: Chrome / Safari


## Screenshots/Videos

https://github.com/Expensify/App/assets/78819774/d28aa3a1-6b7d-4a2f-b88d-e2c7158ab598

**Comment by m-natarajan** [comment]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
## Composer - Edit composer is not focused after canceling The delete confirmation
**Version Number:** 1.4.78-4
**Reproducible in staging?:** N
**Reproducible in production?:** N
**Issue reported by:** Applause-Internal team


## Action Performed:
1. Navigate to a conversation
2. Send a message
3. Click on edit from the action menu.
4. Clear all the texts in the edit composer
5. Click on the send button found on the edit composer
6. Notice that a delete confirmation window is shown. Click on Cancel
## Expected Result:
After canceling, the edit the composer is focused
## Actual Result:
After canceling, the main composer is focused
## Workaround:
Unknown


## Platforms:

Which of our officially supported platforms is this issue occurring on?

- [ ] iOS: Native
- [x] MacOS: Chrome / Safari


## Screenshots/Videos

https://github.com/Expensify/App/assets/38435837/6c152ade-77e8-43e2-be7e-d878cafa146d



</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by kirillzyusko** [comment]
> Do you think you could sync with main

@mountiny done 🙌 

@lanitochka17 @m-natarajan I fixed reported issues as well 👀 (reported by @lanitochka17 indeed was caused by `react-compiler` usage, and reported issue by @m-natarajan was fixed by syncing with latest `main`).

**Comment by mountiny** [comment]
Asking if the testing here as been completed by QA

**Comment by mountiny** [comment]
@kirillzyusko still around 30% of QA left

**Comment by kavimuru** [comment]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
## Status Emoji menu is completely stretched
**Version Number:** 1.4.78-4
**Reproducible in staging?:** N
**Reproducible in production?:** N
**Issue reported by:** Applause-Internal team


## Action Performed:
1. Access https://42287.pr-testing.expensify.com/
2. Sign into a valid account
3. Go to Profile > Status > Emoji dropdown

## Expected Result:
User expects the emoji picker to fit the screen
## Actual Result:
The Emoji picked is completely stretched
## Workaround:
Unknown


## Platforms:

Which of our officially supported platforms is this issue occurring on?

- [ ] iOS: Native
- [x] MacOS: Chrome / Safari


## Screenshots/Videos

https://github.com/Expensify/App/assets/43996225/6b95bbb6-a52f-4282-93aa-9453b953ac89



</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by kirillzyusko** [comment]
> still around 30% of QA left

@mountiny okay, can you ping me when QA finishes all checks so that I can go through the issues and squash all of them out?

**Comment by kirillzyusko** [comment]
But regarding newly reported issues:
- https://github.com/Expensify/App/pull/42287#issuecomment-2178796787 is not reproducible on a fresh build:

> image

- https://github.com/Expensify/App/pull/42287#issuecomment-2178698317

> I think it's common problem that assets are not loading? I'm getting 404 error:
> image
> And I also tested other PR previews (for example this https://github.com/Expensify/App/pull/43993) and it has the same problem. So I tend to think that it's probably BE misconfiguration...

**Comment by lanitochka17** [comment]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
## Settings - UI elements and background are changing position/shape after each click
**Version Number:** 1.4.78-4
**Reproducible in staging?:** N
**Reproducible in production?:** N
**Issue reported by:** Applause-Internal team


## Action Performed:
1. Navigate to https://42287.pr-testing.expensify.com/
2. Navigate to Account settings - Profile
3. Open and close the RHP by clicking on any of the options
4. Navigate to Account settings - Preferences
5. Open and close the RHP by clicking on any of the options
6. Navigate to Account settings - Security
7. Open and close the RHP by clicking on any of the options
8. Navigate to Account settings - Workplaces
9. Click on a workspace
10. Open and close the RHP by clicking on any of the options

## Expected Result:
UI elements and background shouldn't move when an option is clicked

## Actual Result:
UI elements and background are changing position/shape after each click. Affects multiple pages where there are options that open the RHP after interacting

## Workaround:
Unknown


## Platforms:

Which of our officially supported platforms is this issue occurring on?

- [ ] iOS: Native
- [x] MacOS: Chrome / Safari


## Screenshots/Videos

</details> 

https://github.com/Expensify/App/assets/78819774/5c4a7c3f-211f-403b-b47b-ce425b373756


[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by mountiny** [comment]
@lanitochka17 @kavimuru how much do we have left here?

**Comment by kavimuru** [comment]
@mountiny we have 2%left. Will try to finish ASAP.

**Comment by mountiny** [comment]
@kirillzyusko @Ollyws Testing here is done, can you please merge with main and do last testing and verify all the reported issues have been fixed. Thanks!

**Comment by kirillzyusko** [comment]
@mountiny @Ollyws done 🙌

**Comment by mountiny** [comment]
@Ollyws Are you able to prioritize this one today?

**Review by Ollyws — APPROVED** [review]
All reported issues are fixed.

**Review by mountiny — COMMENTED** [review]
@kirillzyusko got one comment

File: .github/actions/javascript/authorChecklist/index.js [hunk 1]
```diff
@@ -21146,7 +21146,8 @@ function TSMappedType(node) {
     nameType,
     optional,
     readonly,
-    typeParameter
+    typeParameter,
```
[the remark this answers is not in the document]
Comment (kirillzyusko) [reply .github/actions/javascript/authorChecklist/index.js:21149]: @mountiny a good question - I don't know why these files have to be re-compiled...

I added only deps to `package.json` and babel plugin to config - can it affect compilation of these files? Or maybe you can explain how these files are compiled (i. e. which configs defines a way how it'll be compiled)?

[the remark this answers is not in the document]
Comment (mountiny) [reply .github/actions/javascript/authorChecklist/index.js:21149]: Yeah it might be due to change from the config. all good then

**Review by mountiny — APPROVED** [review]
Thanks

**Review by rushatgabhane — COMMENTED** [review]
Note: this PR caused a bug - https://github.com/Expensify/App/pull/42287

**Comment by mountiny** [comment]
@rushatgabhane I think you have linked back to this PR

**Pull request merged**